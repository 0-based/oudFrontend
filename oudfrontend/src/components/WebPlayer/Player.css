.img-thumbnail {
  /* max-width: 20%; */
  height: auto;
  /* /* z-index: 4; */
  /* width: 20%; */
  height: 50%;
  bottom: 100px;
  position: fixed;
  padding: 0;
  border: 0;
}
.now-playing-bar-container {
  width: 100%;
  background-color: #383838;
  bottom: 0;
  position: fixed;
  min-width: 620px;
  height: 100px;
  z-index: 4;
}
.now-playing-bar {
  display: flex;
  height: 90px;
  padding: 16px;
  box-sizing: border-box;
}
.now-playing-bar-left,
.now-playing-bar-right {
  width: 30%;
  min-width: 180px;
}
.now-playing-bar-right {
  position: relative;
  margin-top: 16px;
}
.now-playing-bar-left {
  display: flex;
  flex-direction: column;
  align-items: center;
}
.now-playing-bar-left .content .player-controls {
  margin-top: 6px;
}
.now-playing-bar-center {
  width: 40%;
  max-width: 700px;
  display: flex;
  flex-direction: column;
  align-items: center;
}
.now-playing-bar .content {
  width: 100%;
  height: 57px;
}
.now-playing-bar-center .player-controls .control-buttons {
  margin: 0 auto;
  /* display: table; */
  align-self: flex-end;
}
.control-button {
  background-color: transparent;
  border: none;
  vertical-align: middle;
}
.control-button img {
  width: 20px;
  height: 20px;
}
.control-button.play img,
.control-button.pause img {
  width: 32px;
  height: 32px;
  cursor: pointer;
}
.now-playing-bar-left .album-art-work {
  max-width: 66px;
  margin-right: 15px;
  float: left;
  background-size: cover;
  height: 57px;
  width: 66px;
}
.track-controls {
  display: flex;
}
.now-playing-bar-center .track-info {
  align-self: flex-start;
  margin: 0 auto;
  vertical-align: middle;
}
.now-playing-bar-center .track-info .track-name {
  color: #ffce00;
  text-decoration: none;
}
.now-playing-bar-center .track-info .track-name::after {
  content: "  .  ";
}
.now-playing-bar-center .track-info .artist-name {
  font-size: 12px;
  color: #a0a0a0;
  text-decoration: none;
}
a:hover {
  text-decoration: none;
}
button:focus {
  outline: none;
}
.volume-bar {
  width: 180px;
  position: absolute;
  right: 0;
  display: flex;
  align-items: center;
  vertical-align: middle;
}
.ablum-link {
  width: 40%;
  display: block;
}
/* .left-control {
  width: 60%;
  display: flex;
} */
.extended-card-button {
  position: absolute;
  left: 54px;
  top: 18px;
  background: transparent;
  z-index: 1;
  padding: 0;
  display: inline;
  justify-content: center;
  border: 0;
  /* opacity: 0; */
}
.extend-img {
  opacity: 1;
}
.extend-img:hover {
  opacity: 1;
}
.expanded-img {
  width: 20%;
  display: none;
}
.expanded-img-btn {
  position: relative;
  left: 54px;
  top: 18px;
  background: transparent;
  z-index: 4;
  padding: 0;
  display: inline;
  justify-content: center;
  border: 0;
}
.playback-bar {
  display: flex;
}
.progress-time {
  color: #a0a0a0;
  font-size: 11px;
  min-width: 40px;
  text-align: center;
}
.progress-bar {
  width: 100%;
  height: 12px;
  display: inline-flex;
  cursor: pointer;
  background-color: transparent;
}
.playback-bar .progress-bar {
  margin-top: 3px;
}
.progress-bar-bg {
  background-color: #404040;
  height: 4px;
  width: 100%;
  border-radius: 2px;
}
.progress {
  background-color: #ffce00;
  height: 4px;
  width: 0;
  border-radius: 2px;
  transition: 0.1s linear;
  transition-property: width, background-color;
}

@media only screen and (max-width: 615px) {
  .now-playing-bar-container {
    min-width: 320px;
  }
  .ablum-link,
  .previous,
  .next,
  .now-playing-bar-right {
    display: none;
  }
  .play {
    float: left;
    position: absolute;
    left: 80%;
    bottom: 59%;
  }
  .now-playing-bar-center {
    width: 100%;
  }
  .now-playing-bar-left {
    min-width: 0;
    width: 24px;
    height: 24px;
    float: right;
  }
  .now-playing-bar-left .content .player-controls {
    margin: 0;
  }
}
