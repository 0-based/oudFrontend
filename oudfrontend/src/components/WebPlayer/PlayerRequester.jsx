const axios = require("axios");
const provider = "http://localhost:3000/";

export function fetchTrackInfo() {
  let data = null;
  axios
    .get("http://localhost:3000/me/player/currently-playing")
    .then(response => {
      if (response["status"] === 200) {
        data = response;
        console.log("res: " + response.data);
      }
    })
    .catch(function(error) {
      console.log(error);
    });
  return data;
}
