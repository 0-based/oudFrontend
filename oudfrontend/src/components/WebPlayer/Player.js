import React, { Component } from "react";
import Next from "../../assets/images/icons/next.png";
import Pause from "../../assets/images/icons/pause.png";
import Play from "../../assets/images/icons/play.png";
import Previous from "../../assets/images/icons/previous.png";
import Repeat from "../../assets/images/icons/repeat.png";
// import RepeatEnabled from "../../assets/images/icons/repeat-enable.png";
import Shuffle from "../../assets/images/icons/shuffle.png";
// import ShuffleEnabled from "../../assets/images/icons/shuffle-enable.png";
import Volume from "../../assets/images/icons/volume.png";
import art from "../../assets/images/icons/album.jpg";
// import VolumeMuted from "../../assets/images/icons/volume-mute.png";
import "./Player.css";
class WebPlayer extends Component {
  state = {};
  render() {
    return (
      <div className="now-playing-bar-container">
        <div className="now-playing-bar">
          <div className="now-playing-bar-left">
            <div className="content">
              <span className="ablum-link">
                <img src={art} className="album-art-work" alt="Album Art" />
              </span>
              <div className="player-controls">
                <div className="control-buttons">
                  <button className="control-button previous" title="Previous">
                    <img src={Previous} alt="Previous" />
                  </button>
                  <button className="control-button play" title="Play">
                    <img src={Play} alt="Play" />
                  </button>
                  <button
                    className="control-button pause"
                    title="Pause"
                    style={{ display: "none" }}
                  >
                    <img src={Pause} alt="Pause" />
                  </button>
                  <button className="control-button next" title="Next">
                    <img src={Next} alt="Next" />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div className="now-playing-bar-center">
            <div className="content player-controls">
              <div className="track-controls">
                <div className="track-info">
                  <strong className="track-name">Mob Pcycho 2</strong>
                  <strong className="artist-name">Mob</strong>
                </div>

                <div className="control-buttons"></div>
              </div>

              <div className="playback-bar">
                <span className="progress-time current">0.00</span>
                <div className="progress-bar">
                  <div className="progress-bar-bg">
                    <div className="progress"></div>
                  </div>
                </div>
                <span className="progress-time remaining">0.00</span>
              </div>
            </div>
          </div>
          <div className="now-playing-bar-right">
            <div className="volume-bar">
              <button className="control-button shuffle" title="Shuffle">
                <img src={Shuffle} alt="Shuffle" />
              </button>
              <button className="control-button repeat" title="Repeat">
                <img src={Repeat} alt="Repeat" />
              </button>
              <button className="control-button volume" title="Volume">
                <img src={Volume} alt="Volume" />
              </button>
              <div className="progress-bar">
                <div className="progress-bar-bg">
                  <div className="progress"></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default WebPlayer;
