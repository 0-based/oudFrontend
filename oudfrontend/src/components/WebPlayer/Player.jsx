import React, { Component, Fragment } from "react";
import "./Player.css";
import { Howl } from "howler";
import PlayingBarLeft from "./PlayingBarLeft";
import PlayingBarCenter from "./PlayingBarCenter";
import PlayingBarRight from "./PlayingBarRight";
import art from "../../assets/images/icons/album.jpg";
import extend from "../../assets/images/icons/extend.png";
/**
 * Component for playing the audio Oud website, It contains all the player controls.
 * @author Ahmed Ashraf
 * @component
 * @example
 * return (
 *  <Player />
 * )
 */
class Player extends Component {
  /**
   * Setting things up. It initialize what will be fetched from the API
   * @returns{void}
   */
  constructor(props) {
    super(props);
    this.state = {
      sound: null,
      audioUrl: "",
      fetched: false,
      progress: 0,
      playing: false,
      current: Number(0).toFixed(2),
      trackName: "",
      trackId: "",
      artistName: "",
      duration: Number(0).toFixed(2),
      mouseDown: false,
      shuffleState: false,
      repeatState: false,
      volume: 1.0,
      muteState: false,
      thumbHeight: 0,
      thumbDisplay: "initial",
    };
  }
  /**
   * Get the sound progress percentage to update the progress bar
   * @function
   * @returns{number}. returns the progress percentage
   */
  getSoundProgress = () => {
    const progress =
      (this.state.sound.seek() / this.state.sound.duration()) * 100;
    return progress;
  };

  /**
   * Fetching data immediately after the component has been mount to the DOM tree
   */
  componentDidMount() {
    this.fetchPlayback();
  }

  /**
   * Calling the currently playing endpoint to check if there is a track available to be fetched
   * if not, then fetch the recently played track
   * @function
   * @returns{void}
   */
  fetchPlayback = () => {
    console.log("fetchPlayback");
    this.props
      .getRequest("http://localhost:3000/me/player/recently-playing?limit=1")
      .then((response) => {
        const data = response["data"];
        if (!data.hasOwnProperty("status")) {
          const track = data["item"];
          this.setState({
            audioUrl: track["audioUrl"],
            progress: Math.floor(
              (data["progressMs"] / track["duartion"]) * 100
            ),
            playing: data["isPlaying"],
            current: Number(data["progressMs"] / 60000).toFixed(2),
            trackName: track["name"],
            artistName: track["artists"][0]["name"],
            duration: Number(track["duartion"] / 60000).toFixed(2),
            shuffleState: data["shuffleState"],
            repeatState: data["repeatState"] === "off" ? false : true,
            volume: data["device"]["volumePercent"],
            muteState: data["device"]["volumePercent"] === 0 ? true : false,
            fetched: true,
            trackId: track["_id"],
          });
          this.props.changePlayingState(data["isPlaying"]);
          this.props.fetchQueue("0", track["_id"]);
        }
      });
  };

  /**
   * Setting Howler object up.
   * While setting Howler up, It updates the state of the instance to be aware of changes in the UI.
   * Change the playing state to true on playing and false on end of the track
   * Stores the full duration of the track
   * Setting an interval function to update the progress bar periodically
   * On the end of the track return the progress bar to 0
   * @function
   * @returns{void}
   */
  playTrack = () => {
    if (this.state.sound && this.state.sound.state() === "loaded")
      this.state.sound.unload();
    const sound = new Howl({
      src: [this.state.audioUrl],
      autoplay: false,
      loop: this.state.repeatState,
      volume: Number(this.state.volume / 100).toFixed(2),
      mute: this.state.muteState,
      html5: true,
      format: ["mp3"],
      onplay: () => {
        this.props.changePlayingState(true);
        this.setState({
          playing: true,
          duration: Number(this.state.sound.duration() / 60).toFixed(2),
        });
        setInterval(() => {
          if (this.state.sound && this.state.playing) {
            const progress = this.getSoundProgress();
            const current = Number(this.state.sound.seek() / 60).toFixed(2);
            this.setState({
              progress: isNaN(progress) ? this.state.progress : progress,
              current: isNaN(current) ? this.state.current : current,
            });
          }
        }, 100);
      },
      onend: () => {
        this.props.changePlayingState(false);
        this.setState({
          playing: false,
          progress: 0,
          current: Number(0).toFixed(2),
        });
        if (!this.state.repeatState) this.handleNext();
      },
    });
    this.setState({
      sound: sound,
    });
    this.state.sound.play();
    this.state.sound.seek(this.state.current * 60);
    console.log("current: " + this.state.sound.duration());
  };

  /**
   * Handling the pause action. request from the back end to pause, pause the sound in the browser, and update the playing state
   * to false to be aware of the related UI changes
   * @function
   * @returns{void}
   */
  pause = () => {
    this.props
      .putRequest(
        "http://localhost:3000/me/player/pause?deviceId=" + this.props.deviceId
      )
      .then((resp) => {
        this.state.sound.pause();
        this.setState({ playing: false });
        this.props.changePlayingState(false);
      })
      .catch((error) => {
        console.log(error);
      });
  };
  /**
   * PUT request to the server to play/resume the currently track.
   * @function
   * @param {integer} position zero-based index indicates the position of the track in the context array
   * @returns {axios object}
   */
  playResumeRequest = (idx) => {
    return this.props.putRequest(
      "http://localhost:3000/me/player/play?deviceId=" +
        this.props.deviceId +
        "&queueIndex=0",
      { offset: { position: idx } }
    );
  };
  /**
   * Handling the resume action. request from the back end to resume the currently playing track from specific position,
   * resume the sound in the browser, and update the playing state
   * to true to be aware of the related UI changes
   * @function
   * @returns{void}
   */
  resume = (idx = this.props.idx) => {
    this.playResumeRequest(idx)
      .then((resp) => {
        this.props.changePlayingState(true);
        this.setState({ playing: true });
        this.state.sound.play();
        console.log(resp);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  /**
   * Handling the play action. request from the back end to play a track from begining, play the sound in the browser, and update the playing state
   * to true to be aware of the related UI changes
   * @function
   * @returns{void}
   */
  play = (idx = this.props.idx) => {
    this.playResumeRequest(idx)
      .then((resp) => {
        this.playTrack();
        // console.log(resp);
        console.log("play function");
      })
      .catch((error) => {
        console.log(error);
      });
  };

  /**
   * Controlling function for the play, pause, and resume actions. It specifies which action will be called
   * depending on the plaing state of the currently track
   * @function
   * @returns{void}
   */
  handlePlayPause = (id = this.state.trackId, idx = this.props.idx) => {
    if (id !== this.state.trackId) {
      //to uncomment this we need a "REAL" live server
      // this.fetchPlayback();

      this.props
        .fetchTrack(id)
        .then((response) => {
          console.log(response);
          this.setState({
            trackName: response["data"]["name"],
            artistName: response["data"]["artists"][0]["name"],
            audioUrl: response["data"]["audioUrl"],
            duration: response["data"]["duartion"] / 60000,
            current: Number(0).toFixed(2),
            progress: Number(0).toFixed(2),
            playing: true,
            trackId: response["data"]["_id"],
          });
          this.props.changePlayingState(true);
          this.play(idx);
        })
        .catch(function (error) {
          console.log(error);
        });
      return;
    }
    if (this.state.sound) {
      this.state.sound.mute(this.state.muteState);
      this.state.sound.loop(this.state.repeatState);
      if (this.state.sound.playing()) {
        this.pause();
      } else if (this.state.sound.state() === "loaded") {
        this.resume(idx);
      }
    } else this.play(idx);
  };

  /**
   * Handling the next action.
   * Request from the server to get the next track then play it
   * @function
   * @returns{void}
   */
  handleNext = () => {
    this.props
      .postRequest(
        "http://localhost:3000/me/player/next?deviceId=" + this.props.deviceId
      )
      .then((response) => {
        this.props
          .getNext()
          .then((response) => {
            console.log(response);
            this.setState({
              trackName: response["data"]["name"],
              artistName: response["data"]["artists"][0]["name"],
              audioUrl: response["data"]["audioUrl"],
              duration: response["data"]["duartion"] / 60000,
              current: Number(0).toFixed(2),
              progress: Number(0).toFixed(2),
              playing: false,
              trackId: response["data"]["_id"],
            });
            this.props.changePlayingState(false);
            this.play();
          })
          .catch(function (error) {
            console.log(error);
          });
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  /**
   * Handling the previous action.
   * Request from the server to get the previous track then play it
   * @function
   * @returns{void}
   */
  handlePrev = () => {
    this.props
      .postRequest(
        "http://localhost:3000/me/player/previous?deviceId=" +
          this.props.deviceId
      )
      .then((response) => {
        this.props
          .getPrevious()
          .then((response) => {
            this.setState({
              trackName: response["data"]["name"],
              artistName: response["data"]["artists"][0]["name"],
              audioUrl: response["data"]["audioUrl"],
              duration: response["data"]["duartion"] / 60000,
              current: Number(0).toFixed(2),
              progress: Number(0).toFixed(2),
              playing: false,
              trackId: response["data"]["_id"],
            });
            this.props.changePlayingState(false);
            this.play();
          })
          .catch(function (error) {
            console.log(error);
          });
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  /**
   * Handling the event of clicking on the progress bar using the mouse
   * It gets the mouse click position then normalize it to the width of the bar
   * then request from the server to progress to the specified position and update the state with new progress
   * @function
   * @returns{void}
   */
  onProgressClick = (e) => {
    // e.preventDefault();
    if (!this.state.mouseDown || !this.state.sound) return;
    const width = document.getElementById("progress-width").clientWidth;
    const offsetX = e.nativeEvent.offsetX;
    // const offsetWidth = e.nativeEvent.target.offsetWidth;
    const percent = offsetX / width;
    const position = percent * this.state.duration * 60;

    this.props
      .putRequest(
        "http://localhost:3000/me/player/seek?deviceId=" + this.props.deviceId
      )
      .then((response) => {
        this.state.sound.seek(position);
        this.setState({
          progress: this.getSoundProgress(),
        });
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  /**
   * Update the mouse down flag to be aware of the mouse click on the progress or volume bar
   * @function
   * @returns{void}
   */
  setMouseDown = (cond) => {
    this.setState({
      mouseDown: cond,
    });
  };

  /**
   * Handling the shuffle action.
   * Request from the server to shuffle the current queue then update the shuffle state to render the proper button
   * @function
   * @returns{void}
   */
  handleShuffleState = () => {
    this.props
      .putRequest(
        "http://localhost:3000/me/player/shuffle?deviceId=" +
          this.props.deviceId
      )
      .then((response) => {
        this.setState({
          shuffleState: !this.state.shuffleState,
        });
        console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  /**
   * Handling the repeat action.
   * Request from the server to repeat the currently playing track then update the repeat state to render the proper button
   * then active the repeat state in Howler
   * @function
   * @returns{void}
   */
  handleRepeatState = () => {
    this.props
      .putRequest(
        "http://localhost:3000/me/player/repeat?deviceId=" + this.props.deviceId
      )
      .then((response) => {
        const loop = !this.state.repeatState;
        this.setState({
          repeatState: loop,
        });
        if (this.state.sound) this.state.sound.loop(loop);
        console.log("from repeat it's fine!");
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  volumeRequest = (volumePercent = this.state.volume) => {
    return this.props.putRequest(
      "http://localhost:3000/me/player/volume?deviceId=" + this.props.deviceId
    );
    //+"&volumePercent=" + volumePercent
  };
  /**
   * Handling the mute action.
   * Request from the server to mute the currently playing track then update the mute state to render the proper button
   * then active the mute state in Howler
   * @function
   * @returns{void}
   */
  handleMuteState = () => {
    const mute = !this.state.muteState;
    this.volumeRequest(0)
      .then((response) => {
        this.setState({
          muteState: mute,
        });
        if (this.state.sound) this.state.sound.mute(mute);
        console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  /**
   * Handling the event of clicking on the volume bar using the mouse
   * It gets the mouse click position then normalize it to the width of the bar
   * then request from the server to progress to the specified volume and update the state with new volume
   * @function
   * @returns{void}
   */
  onVolumeClick = (e) => {
    // e.preventDefault();
    if (!this.state.mouseDown || !this.state.sound) return;
    const width = document.getElementById("volume-width").clientWidth;
    const offsetX = e.nativeEvent.offsetX;
    const percent = offsetX / width;
    const volume = parseInt(percent * 100);

    this.volumeRequest(volume)
      .then((response) => {
        this.state.sound.volume(volume / 100);
        this.setState({
          volume: volume,
          muteState: volume === 0 ? true : false,
        });
        console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
  };
  closeThumb = () => {
    this.setState({
      thumbHeight: 0,
      thumbDisplay: "initial",
    });
  };
  openThumb = () => {
    this.setState({
      thumbHeight: "35%",
      thumbDisplay: "none",
    });
  };
  render() {
    return (
      <Fragment>
        <div
          className="extedned-thumb"
          style={{ height: this.state.thumbHeight }}
        >
          <a href="/">
            <img src={art} alt="Cinque Terre" className="thumb-img" />
          </a>
          <button className="close-thumb" onClick={this.closeThumb}>
            <img src={extend} alt="Close Queue" />
          </button>
        </div>

        <div className="now-playing-bar-container" data-testid="web-palyer">
          <div className="now-playing-bar">
            <PlayingBarLeft
              playing={this.state.playing}
              display={this.state.thumbDisplay}
              handlePrev={() => this.handlePrev()}
              handlePlayPause={() => this.handlePlayPause()}
              handleNext={() => this.handleNext()}
              openThumb={this.openThumb}
              data-testid="web-player-left"
            />

            <PlayingBarCenter
              trackName={this.state.trackName}
              artistName={this.state.artistName}
              current={this.state.current}
              progress={this.state.progress}
              duration={this.state.duration}
              setMouseDown={() => this.setMouseDown(true)}
              onProgressClick={(e) => this.onProgressClick(e)}
              mouseUp={(e) => {
                this.onProgressClick(e);
                document.addEventListener(
                  "mouseup",
                  this.setState({
                    mouseDown: false,
                  })
                );
              }}
              data-testid="web-player-center"
            />

            <PlayingBarRight
              shuffleState={this.state.shuffleState}
              repeatState={this.state.repeatState}
              volumeState={this.state.muteState}
              volume={this.state.volume}
              queueElement={this.props.queueElement}
              handleShuffleState={this.handleShuffleState}
              handleRepeatState={this.handleRepeatState}
              handleMuteState={this.handleMuteState}
              setMouseDown={() => this.setMouseDown(true)}
              onVolumeClick={(e) => this.onVolumeClick(e)}
              mouseUp={(e) => {
                this.onVolumeClick(e);
                document.addEventListener(
                  "mouseup",
                  this.setState({
                    mouseDown: false,
                  })
                );
              }}
              data-testid="web-player-right"
            />
          </div>
        </div>
      </Fragment>
    );
  }
}

export default Player;
