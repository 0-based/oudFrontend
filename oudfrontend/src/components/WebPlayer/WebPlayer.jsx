import React, { Fragment } from "react";
import { Component } from "react";
import axios from "axios";
import Player from "./Player/Player";
import Queue from "./Queue/Queue";
import Swal from "sweetalert2";
import { saveTrack, removeSavedTrack } from "../../utils/Actions/Player";
const config = {
  headers: {
    authorization: `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlOTA3ZGFmYTA2NDVmNDU3MTYwNzVmZiIsImlhdCI6MTU4Njg5MzE0MywiZXhwIjoxNTg5NDg1MTQzfQ.ON2Ef2vgOV1_6EokwvD3mlUzgAn0pb5WPCy5qBWj2QA`,
  },
};
/**
 * Component controlling the Player and Qeueu Components. It masters all the logic behind the web player.
 * @author Ahmed Ashraf
 * @component
 * @example
 * return (
 *  <WebPlayer />
 * )
 */
class WebPlayer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      context: {},
      trackIdx: 0,
      trackId: "",
      queue: [],
      deviceId: "74ASZWbe4lXaubB36ztrGX",
      playing: false,
      loved: false,
    };
    this.queueElement = React.createRef();
    this.playerElement = React.createRef();
  }
  /**
   * Axios GET request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @returns {object}
   */
  getRequest = (endpoint) => {
    return axios.get(endpoint, config);
  };
  /**
   * Axios DELETE request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @returns {object}
   */
  deleteRequest = (endpoint) => {
    return axios.delete(endpoint, {}, config);
  };
  /**
   * Axios PUT request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  putRequest = (endpoint, body = {}) => {
    return axios.put(endpoint, body, config);
  };
  /**
   * Axios POST request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  postRequest = (endpoint, body = {}) => {
    return axios.post(endpoint, body, config);
  };
  /**
   * Axios PATCH request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  patchRequest = (endpoint, body = {}) => {
    return axios.patch(endpoint, body, config);
  };
  /**
   * A function to fetch the queue immediately afer the player has mounted.
   * @function
   * @param {number} queueIndex The index of the queue Default 0 must be greater than 0 if 0 get the current queue if 1 get the last queue
   * @param {number} trackId the id of the currently fetched track in the Player component
   * @returns {void}
   */
  fetchQueue = (queueIndex = "0", trackId = "", newQueue = false) => {
    this.getRequest(
      "https://oud-zerobase.me/api/v1/me/queue?queueIndex=" + queueIndex
    )
      .then((response) => {
        const data = response["data"];
        if (!data.hasOwnProperty("status")) {
          console.log("queeu: ");
          console.log(response);
          const size = data["total"];
          let trackIdx = 0;
          for (let i = 0; i < size; ++i)
            if (data["tracks"][i] === trackId) {
              console.log("found track id");
              trackIdx = i;
              break;
            }

          console.log("track id from data: " + data["tracks"][0]);
          console.log("track id from params: " + trackId);
          this.setState({
            trackIdx: newQueue ? 0 : trackIdx,
            queue: data["tracks"],
            trackId: newQueue ? data["tracks"][0] : trackId,
          });
          console.log("track id from state: " + this.state.trackId);
        }
      })
      .catch(function (error) {
        console.log(error.response.data.message);
      });
  };
  /**
   * A function to fetch any track from the server
   * @function
   * @param {number} trackId the id of the wanted track
   * @returns {object}
   */
  fetchTrack = (trackId) => {
    let trackIdx = this.state.trackIdx;
    for (let i = 0; i < this.state.queue.length; ++i)
      if (this.state.queue[i] === trackId) {
        trackIdx = i;
        break;
      }
    this.setState({
      trackIdx: trackIdx,
    });
    return this.getRequest("https://oud-zerobase.me/api/v1/tracks/" + trackId);
  };
  /**
   * A function to fetch the next track to the currently playing track
   * @function
   * @returns {object}
   */
  getNext = () => {
    let idx = this.state.trackIdx + 1;
    if (idx && this.playerElement.current.state.repeatState === 1) {
      idx = 0;
    } else idx = idx - 1;

    this.setState({
      trackIdx: idx,
      trackId: this.state.queue[idx],
    });
    return this.state.trackId;
  };
  /**
   * A function to fetch the previos track to the currently playing track
   * @function
   * @returns {object}
   */
  getPrevious = () => {
    const idx = this.state.trackIdx - 1 < 0 ? 0 : this.state.trackIdx - 1;
    this.setState({
      trackIdx: idx,
      trackId: this.state.queue[idx],
    });
    return this.state.trackId;
  };
  /**
   * A generic function to play a track from any location other than the queue and the player
   * @function
   * @param {string} context OUD context
   * @param {array} uris OUD uris
   * @param {number} offset indext of the track in the context
   * @param {number} position from where to start in the track. In milliseconds
   * @reutrns {void}
   */
  playContext = (
    contextUri = this.state.context,
    uris = [],
    offset = 0,
    position = 0
  ) => {
    this.putRequest(
      "https://oud-zerobase.me/api/v1/me/player/play?deviceId=" +
        this.state.deviceId +
        "&queueIndex=0",
      {
        contextUri: contextUri,
        uris: uris,
        offset: { position: offset },
        positionMs: position,
      }
    )
      .then((response) => {
        // this.fetchQueue();
        let player = this.playerElement.current;
        player.fetchPlayback(true);
        setTimeout(() => {
          player.playTrack();
        }, 100);

        // if (
        //   response["data"].hasOwnProperty("status") &&
        //   response["data"]["status"] === "204"
        // ) {
        //   console.log("new context has fetched");
        //   this.fetchQueue();
        //   this.playerElement.fetchPlayback();
        //   this.playerElement.playTrack();
        // } else console.log(response);
      })
      .catch((error) => {
        console.log(error.response.data.message);
      });
  };
  /**
   * Handling the change in the order of the tracks in the queue.
   * @function
   * @param {array} queue the currently playing queue
   * @returns {void}
   */
  onChangeQueueOrder = (
    oldIdx = this.state.trackIdx,
    newIdx = this.state.trackIdx
  ) => {
    console.log("old index: " + oldIdx);
    console.log("new index: " + newIdx);
    this.patchRequest(
      "https://oud-zerobase.me/api/v1/me/queue?queueIndex=0" +
        "&trackIndex=" +
        oldIdx +
        "&newIndex=" +
        newIdx
    )
      .then((response) => {
        this.fetchQueue(0, this.state.trackId);
        console.log("changed order");
        console.log(response);
      })
      .catch((error) => {
        console.log(error.response);
        console.log("error changing order");
      });
  };
  /**
   * Change the playing state as there are many stuff depend on that in the UI
   * @function
   * @param {boot} playing playing state of the track
   * @returns {void}
   */
  changePlayingState = (
    playing,
    idx = this.state.trackIdx,
    id = this.state.trackId
  ) => {
    this.setState({
      playing: playing,
      trackIdx: idx,
      trackId: id,
    });
  };
  togglePlayingState = () => {
    this.setState({
      playing: !this.state.playing,
    });

    return this.state.playing;
  };
  changeLovedState = (state) => {
    this.setState({
      loved: state,
    });
  };
  /**
   * Hadling the removing of a track from the queue
   * @function
   * @param {number} idx the index of the track in the queue
   * @param {string} id the unique id of the track
   * @returns {void}
   */
  removeTrack = (idx, id) => {
    //
    this.deleteRequest("https://oud-zerobase.me/api/v1/me/queue?trackId=" + id)
      .then((response) => {
        console.log(response);
        this.fetchQueue(0, this.state.trackId);
        Swal.fire({
          title: "Done!",
          text: "Track Deleted Successfully from your Queue!",
          icon: "success",
          showConfirmButton: false,
          timer: 1000,
        });
      })
      .catch((error) => {
        console.log(error.response.data.message);
      });
  };
  likeSong = (trackId, queue = false) => {
    saveTrack(trackId).then((done) => {
      if (done) {
        this.setState({
          loved: queue ? this.state.loved : true,
        });
        Swal.fire({
          title: "Done!",
          text: "Added to your Liked Songs!",
          icon: "success",
          showConfirmButton: false,
          timer: 1000,
        });
      }
    });
  };
  unlikeSong = (trackId, queue = false) => {
    removeSavedTrack(trackId).then((done) => {
      if (done)
        this.setState({
          loved: queue ? this.state.loved : false,
        });
      Swal.fire({
        title: "Done!",
        text: "Removed from your Liked Songs!",
        icon: "success",
        showConfirmButton: false,
        timer: 1000,
      });
    });
  };
  render() {
    return (
      <Fragment>
        <Queue
          ref={this.queueElement}
          tracks={this.state.queue}
          trackId={this.state.trackId}
          trackIdx={this.state.trackIdx}
          deviceId={this.state.deviceId}
          onChangeQueueOrder={this.onChangeQueueOrder}
          player={this.playerElement}
          removeTrack={this.removeTrack}
          likeSong={this.likeSong}
          unlikeSong={this.unlikeSong}
          data-testid="queue-container"
        />
        <Player
          ref={this.playerElement}
          deviceId={this.state.deviceId}
          trackIdx={this.state.trackIdx}
          queueElement={this.queueElement}
          getRequest={this.getRequest}
          putRequest={this.putRequest}
          postRequest={this.postRequest}
          fetchQueue={this.fetchQueue}
          getNext={this.getNext}
          getPrevious={this.getPrevious}
          changePlayingState={this.changePlayingState}
          fetchTrack={this.fetchTrack}
          likeSong={this.likeSong}
          unlikeSong={this.unlikeSong}
          loved={this.state.loved}
          changeLovedState={this.changeLovedState}
          data-testid="player-container"
        />
      </Fragment>
    );
  }
}

export default WebPlayer;
