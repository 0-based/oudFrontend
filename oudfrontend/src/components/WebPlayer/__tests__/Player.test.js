import React from "react";
import WebPlayer from "../Player";
import renderer from "react-test-renderer";
// import { cleanup } from "@testing-library/react";
import Enzyme, { shallow, mount } from "enzyme";

//testing whether the WebPlayer component renders.
describe("Web Player component", () => {
  test("renders", () => {
    const wrapper = shallow(<WebPlayer />);
    expect(wrapper.exists()).toBe(true);
  });
  test("snap shot", () => {
    const wrapper = shallow(<WebPlayer />);
    expect(wrapper).toMatchSnapshot();
  });
});
