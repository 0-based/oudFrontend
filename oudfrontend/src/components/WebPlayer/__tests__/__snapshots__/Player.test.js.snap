// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Player component snap shot 1`] = `ShallowWrapper {}`;
